### PR TITLE
[sailfish-browser] Display a handle at the top of the popup menu. Fixes JB#54528 OMP#JOLLA-183

### PR DESCRIPTION
**Pull request by adenexter**

### CODE REVIEW

**Comment by rainemak** [comment]
I think we should review the rest as one meaning that we should "ignore" https://github.com/sailfishos/sailfish-browser/pull/857

**Comment by llewelld** [comment]
This is really nice behaviour. I'm experiencing a bit of a glitch though, when a page is added to or removed from the pageStack. The horizontal offset seems to mix badly with the shader positioning. These screenshots are from selecting the History item in the menu.

menuslide

**Comment by adenexter** [comment]
> This is really nice behaviour. I'm experiencing a bit of a glitch though

I commented here first, https://github.com/sailfishos/sailfish-browser/pull/856#discussion_r649624624

I guess I will have to use a custom shader for the background after all in order to change the reference coordinates for the texure coordinates. It should be trivial but I just could not make it work the other day.

**Comment by adenexter** [comment]
>  It should be trivial but I just could not make it work the other day.

Now I know why I couldn't make it work. The scene graph renderer pre-transforms geometry so it can batch together different items into a single render call. So my single item which I create with the geometry (0, 0, 768x1419) gets submitted to the GPU as (200, 32, 768x1419) or something worse if the item is rotated or otherwise transformed. All of which means I can't normalize the position using the item size to get texture coordinates which is what I was trying to do.

So I can add a normalized position coordinates to the Background item's geometry, which I already wanted for something before but also ran into odd issues with at the time. Or do a limited fix by counter transforming some items based on page coordinates, which I'm already realizing won't work because of animator transitions.

**Comment by rainemak** [comment]
> > It should be trivial but I just could not make it work the other day.
> So I can add a normalized position coordinates to the Background item's geometry, which I already wanted for something before but also ran into odd issues with at the time. Or do a limited fix by counter transforming some items based on page coordinates, which I'm already realizing won't work because of animator transitions.

Idea / thought: How would it look like if we'd throw away that Background and ShaderEffectSource and just have a Rectangle with a rounded corners (~matching coloring) + a little bit transparency.

**Comment by adenexter** [comment]
> Idea / thought: How would it look like if we'd throw away that Background and ShaderEffectSource and just have a Rectangle with a rounded corners (~matching coloring) + a little bit transparency.

The scroll bar and item press highlights won't be clipped that way. And the footer still has a Background item or a large rectangle clipped so only its bottom corners are rounded.

I've found my problem now, it requires a change to Silica but it's something I've wanted before and probably will again.

File: apps/browser/qml/pages/components/PopUpMenu.qml [hunk 1]
```diff
@@ -30,74 +30,112 @@ SilicaControl {
...
+                    id: headerItem
+
+                    y: Math.max(0, footerLoader.y - height - (Screen.sizeCategory > Screen.Medium
+                                ? contentLoader.height
+                                : Math.min(contentLoader.height, Theme.paddingLarge * popUpMenu.heightRatio)))
+
+                    width: menuFlickable.width
+                    height: Theme.paddingLarge
+
+                    Rectangle {
```
Comment (rainemak) [inline apps/browser/qml/pages/components/PopUpMenu.qml:105]: Let's raise this above actual content items. On landscape menu expands to the max position leaving only panning of the flickable. When panning menu downwards, you can see all menu items going above the handle

File: apps/browser/qml/pages/components/PopUpMenu.qml [hunk 2]
```diff
@@ -30,74 +30,112 @@ SilicaControl {
...
         }
     }
 
-    FadeAnimation {
-        id: menuFadeAnimation
+    NumberAnimation {
+        id: menuAnimation
 
         running: false
+        duration: 2000
```
Comment (rainemak) [inline apps/browser/qml/pages/components/PopUpMenu.qml:58]: This feels now like a leftover from the testing.

Comment (rainemak) [inline apps/browser/qml/pages/components/PopUpMenu.qml:58]:
```suggestion
        duration: 200
```

Comment (adenexter) [reply apps/browser/qml/pages/components/PopUpMenu.qml]: Done

File: apps/browser/qml/pages/components/PopUpMenu.qml (hunk 1 above)
Comment (adenexter) [reply apps/browser/qml/pages/components/PopUpMenu.qml]: Done if I've understood correctly.

**Comment by adenexter** [comment]
> I've found my problem now, it requires a change to Silica but it's something I've wanted before and probably will again.

A fix for this has been merged to Silica.

File: apps/browser/qml/pages/components/PopUpMenu.qml [hunk 3]
```diff
@@ -117,22 +140,42 @@ SilicaControl {
                             flickable: menuFlickable
                         }
                     },
+                    Rectangle {
```
Comment (llewelld) [inline apps/browser/qml/pages/components/PopUpMenu.qml:143]: With the new variable dimensions of `menuFlickable` I'm seeing some strange behaviour from the `VerticalScrollDecorator`. In practice it seems the values you need going in to the decorator are:
```
flickable.Height: menuFlickable.height - menuItem.menuTop - (2 * popUpMenu.margin)
flickable.contentHeight: menuFlickable.contentHeight - contentLoader.y
flickable.contentY: menuFlickable.contentY + menuItem.menuTop - contentLoader.y + (2 * popUpMenu.margin)
```
In other words, giving:
```
_sizeRatio: ((menuFlickable.height - menuItem.menuTop - (2 * popUpMenu.margin)) - _headerSpacing) / (menuFlickable.contentHeight - contentLoader.y + _topMenuSpacing + _bottomMenuSpacing)
y: Math.max(0, Math.min((parent.height / (menuFlickable.height - menuItem.menuTop - (2 * popUpMenu.margin))) * (_headerSpacing + ((menuFlickable.contentY + menuIt    em.menuTop - contentLoader.y + (2 * popUpMenu.margin)) - flickable.originY + _topMenuSpacing) * _sizeRatio), (parent.height - height)))
```

**Review by llewelld — CHANGES_REQUESTED** [review]
This is working nicely, no artefacts when the menu slides on/off the page in any direction. There's some strange behaviour from the scroll decorator though (a minor thing, see my comment elsewhere).

File: apps/browser/qml/pages/components/PopUpMenu.qml (hunk 3 above)
Comment (adenexter) [reply apps/browser/qml/pages/components/PopUpMenu.qml:152]: Done, in a simplified fashion. The page header and other margins in the source item are things we explicitly don't want as the presence of a dialog header in a Page containing the PopUpMenu should not have any effect on the menu's scroll decorator.

Comment (llewelld) [reply apps/browser/qml/pages/components/PopUpMenu.qml:152]: Okay, perfect, this is working nicely now. Thank you.

**Review by llewelld — APPROVED** [review]
This is working nicely now for me on an XA2. The whole menu approach does now seem quite clear and usable (it's clear if there are more elements in the menu, there's good use of the screen space, good alignment with the other items on the toolbar, etc.).

**Review by rainemak — APPROVED** [review]
LGTM as well.

**Review by atatarov — APPROVED** [review]
Thank you! I guess we don't have the solution for the tablet, but it would be nice to have ability to close "popup menu" by swiping from top to bottom :)

**Comment by rainemak** [comment]
> Thank you! I guess we don't have the solution for the tablet, but it would be nice to have ability to close "popup menu" by swiping from top to bottom :)

I think there are two improvements that we should consider but let's handle both of them separately

1) Dimming the content area when popup menu like we do elsewhere => this should help users to understand that tapping outside of the popup menu will dismiss it

2) Swipe to close gesture for the popup menu

**Comment by okodron** [comment]
> > Thank you! I guess we don't have the solution for the tablet, but it would be nice to have ability to close "popup menu" by swiping from top to bottom :)
> 
> I think there are two improvements that we should consider but let's handle both of them separately
> 
>     1. Dimming the content area when popup menu like we do elsewhere => this should help users to understand that tapping outside of the popup menu will dismiss it
> 
>     2. Swipe to close gesture for the popup menu

Yes, good idea. Better to file a separate task about this so that it won't be forgotten.

**Review by aroditelev — APPROVED** [review]

**Pull request merged**